Caching, retry and de-duplication of incoming messages

scottnonnenberg opened the pull request:
When a message is first received, we cache it. After we've decrypted it, we update that cached copy of the message. The message is only removed from our cache when the application confirms that it is done processing it. On startup we load all messages from the cache and attempt to process them once more. We'll try five times before giving up.

Additionally, for messages visible to the user, we check for duplicates before letting the application process any incoming message.

Architecture:
- Two new migrations. The first is for `unprocessed` messages, with one currently-unused index by `timestamp` (when the message was received). The second is to add a `unique` index to the `messages` data store on three fields `[source, sourceDevice, sent_at]`
- All events (except `error`) emitted by `MessageReceiver` now include a `confirm` property, a function which will remove the original incoming envelope from the cache. This is expected to be called after the application has fully processed the message.
- We save both the original `envelope` and the fully-decrypted message because our protocol layer is stateful. If we decrypt a message, we throw away the key used to decrypt that message. Attempting to decrypt it again will cause an error.

Trolldemorted commented:
I suppose ff259d80746a0ee470a7f0837290bf177043e5ff fixes https://github.com/WhisperSystems/Signal-Desktop/issues/1038 - can't wait to see this live! :+1:

liliakai commented on js/background.js:
```diff
@@ -179,14 +179,24 @@
-        message.handleDataMessage(data.message);
+        var message = initIncomingMessage(data);
+
+        isMessageDuplicate(message).then(function(isDuplicate) {
```
Might we also want to do this in onSentMessage?

scottnonnenberg replied:
Yup! Will be pushing a commit very soon.

liliakai commented on js/background.js:
```diff
@@ -341,7 +377,9 @@
+            timestamp: timestamp,
+            source: pushMessage.source,
+            confirm: ev.confirm
         });
```
Maybe use the 'remove' event on the receipt instead of passing through the confirm method?
```
var receipt = Whipser.DeliveryReceipts.add({...});
receipt.on('remove', confirm);
```

scottnonnenberg replied:
Done!

liliakai commented on js/views/conversation_view.js:
```diff
@@ -366,15 +366,17 @@
-                this.scrollDownButton.remove();
+                var button = this.scrollDownButton;
                 this.scrollDownButton = null;
+                button.remove();
```
Against what are you defending, exactly?

scottnonnenberg replied:
Not exactly sure. Received a report that the last seen indicator didn't go away on send. This will ensure that if the `.remove()` crashes, we'll still set `this.lastSeenIndicator = null`.

scottnonnenberg commented on js/database.js:
```diff
@@ -254,6 +254,26 @@
+                console.log('migration 14.0');
+                console.log('Adding unprocessed message store');
+                var unprocessed = transaction.db.createObjectStore('unprocessed');
+                unprocessed.createIndex('received', ['timestamp'], { unique: false });
```
There probably is, but we don't want it. :0) Commit arriving shortly.

scottnonnenberg commented:
Think I've addressed all the feedback. That last commit,  https://github.com/WhisperSystems/Signal-Desktop/pull/1314/commits/f33e7802790a6ca9b9e3b2371e9239f0f7f7e830, breaks the dependency `libtextsecure.js` -> `Whisper.Unprocessed`, moving to a mechanism like the rest of `signal_protocol_store.js`.

liliakai commented on libtextsecure/message_receiver.js:
```diff
@@ -288,7 +267,7 @@ MessageReceiver.prototype.extend({
         }
         return promise.then(function(plaintext) {
-            this.updateCached(envelope, plaintext);
+            this.updateCache(envelope, plaintext);
```
Not blocking on the update here. Is that ok?

liliakai commented on js/signal_protocol_store.js:
```diff
@@ -773,6 +782,43 @@
+                // Return a plain array of plain objects
+                return collection.map(function(unprocessed) {
+                    return unprocessed.attributes;
+                });
```
BackboneCollection map offers some syntactic sugar: `return collection.map('attributes');`

liliakai commented on libtextsecure/message_receiver.js:
```diff
@@ -154,42 +126,49 @@ MessageReceiver.prototype.extend({
+                }
+            }.bind(this));
+
+            return items;
```
Maybe we removed some items from storage or updated them, but we still return them without blocking on the write to storage. Is it safe? Might we cause a race between remove/update and an update after decryption? Possibly the queuing in queueCache ameliorates that danger, so just something to be mindful of.

scottnonnenberg commented:
Updated the location of `this.remove()` for both read and delivery receipts. Next I'm gonna spend some bit of time to add testing for the stuff dealing with `Unprocessed` models as well as `MessageReceiver`.
